graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -388,8 +541,170 @@ def enter_field_definition(
+        """Create a visitor to check that suppression does not cause an illegal state.
+
+        Args:
+            renamings: Dict[str, Optional[str]], from original field name to renamed field name or
+                       None (for type suppression). Any name not in the dict will be unchanged
+            query_type: str, name of the query type (e.g. RootSchemaQuery)
+        """
+        self.renamings = renamings
+        self.query_type = query_type
+        self.current_type: Optional[str] = None
+        # Maps a type T to a dict which maps a field F belonging to T to the field's type T'
+        self.fields_to_suppress: Dict[str, Dict[str, str]] = {}
+        # Record any unions to suppress because all their types were suppressed
+        self.union_types_to_suppress: List[UnionTypeDefinitionNode] = []
+
+    def enter_object_type_definition(
+        self,
+        node: ObjectTypeDefinitionNode,
+        key: Any,
+        parent: Any,
+        path: List[Any],
+        ancestors: List[Any],
+    ) -> None:
+        """If the node's name matches the query type, record that we entered the query type."""
+        self.current_type = node.name.value
+
+    def leave_object_type_definition(
+        self,
+        node: ObjectTypeDefinitionNode,
+        key: Any,
+        parent: Any,
+        path: List[Any],
+        ancestors: List[Any],
+    ) -> None:
+        """If the node's name matches the query type, record that we left the query type."""
+        self.current_type = None
+
+    def enter_field_definition(
+        self,
+        node: FieldDefinitionNode,
+        key: Any,
+        parent: Any,
+        path: List[Any],
+        ancestors: List[Any],
+    ) -> None:
+        """If not at query type, check that no field depends on a type that was suppressed."""
+        if self.current_type == self.query_type:
+            return None
+        # At a field of a type that is not the query type
+        field_name = node.name.value
+        field_type = get_ast_with_non_null_and_list_stripped(node.type).name.value
+        if self.renamings.get(field_type, field_type):
+            return None
+        # Reaching this point means this field is of a type to be suppressed.
+        if self.current_type is None:
+            raise AssertionError(
+                "Entered a field not in any ObjectTypeDefinition scope because "
+                "self.current_type is None"
+            )
+        if self.current_type == field_type:
+            # Then node corresponds to a field belonging to type T that is also of type T.
+            # Therefore, we don't need to explicitly suppress the field as well and this should not
+            # raise errors.
+            return None
+        if self.current_type not in self.fields_to_suppress:
+            self.fields_to_suppress[self.current_type] = {}
+        self.fields_to_suppress[self.current_type][field_name] = field_type
```

Comment:
You could also do `self.fields_to_suppress.setdefault(self.current_type, {})[field_name] = field_type` to do this as a one-liner. No particular preference for it one way or the other, this is fine.